kvcoord: elide locking requests in txnWriteBuffer

stevendanna opened the pull request:
If we know that a sufficient lock is already held at the given sequence number, we can often elide duplicate locking requests. Currently, the only requests that modify the buffer are write requests.

The test adds skipped cases that would require us to buffer the returned locks and values.

Epic: none
Release note: None

stevendanna committed:
kvcoord: elide locking requests in txnWriteBuffer

If we know that a sufficient lock is already held at the given
sequence number, we can often elide duplicate locking
requests. Currently, the only requests that modify the buffer are
write requests.

The test adds skipped cases that would require us to buffer the returned locks
and values.

Informs #143896

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/148438)
<!-- Reviewable:end -->

stevendanna committed:
kvpb: allow CanForwardReadTimestamp and ExpectExclusionSince for RC

CanForwardReadTimestamp and ExpectExclusionSince can co-exist in
transactions that use per statement read snapshots since the read
footprint gets reset each time the transaction is stepped.

Epic: none
Release note: None

yuzefovich reviewed (commented):
Nice work! I have a few questions.

Also should this PR close #143896? It definitely should be linked to that issue, not sure if there is more follow-up.

> Reviewed 3 of 3 files at r2, 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/148438#-:-OT-VPBg3YovCcLcBOaV:b-s51ozx)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @stevendanna)

___
*[`pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer.go` line 805 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/148438#-OT-QLJKERKyHGY0--mA:-OT-QLJKERKyHGY0--mB:b15q855) ([raw file](https://github.com/cockroachdb/cockroach/blob/f977a956c52efd38858152173cb8b7b88a276dd8/pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer.go#L805)):*
> ```Go
> 			record.stripped = !t.MustAcquireExclusiveLock
> 			record.transformed = t.MustAcquireExclusiveLock
> ```

nit: let's keep the empty line for symmetry.
___
*[`pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer_test.go` line 2795 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/148438#-OT-W1XG2XsvfyVUSmbo:-OT-W1XG2XsvfyVUSmbp:beduld4) ([raw file](https://github.com/cockroachdb/cockroach/blob/47ec2e06784b20fd12ea6ea77562b4c397e42983/pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer_test.go#L2795)):*
> ```Go
> 
> 		lki = newLockedKeyInfo(lock.Shared, 2, ts1)
> 		lki.acquireLock(lock.Exclusive, 2, ts1)
> ```

nit: maybe a slightly more interesting test case would be acquire Shared at 2 and then acquire Exclusive at 3.
___
*[`pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer_test.go` line 2851 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/148438#-OT-WBwv0KlOU5hvrvSP:-OT-WBwv0KlOU5hvrvSQ:bjuv80f) ([raw file](https://github.com/cockroachdb/cockroach/blob/47ec2e06784b20fd12ea6ea77562b4c397e42983/pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer_test.go#L2851)):*
> ```Go
> 		name string
> 
> 		// requiresValue, when true, indicates that this request requires.
> ```

nit: seems incomplete.
___
*[`pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer_test.go` line 2864 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/148438#-OT-WJXyB4q1FBb7pAvb:-OT-WJXyB4q1FBb7pAvc:b-m5e8j4) ([raw file](https://github.com/cockroachdb/cockroach/blob/47ec2e06784b20fd12ea6ea77562b4c397e42983/pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer_test.go#L2864)):*
> ```Go
> 		// completely from the buffer if a previous request locked the value. This
> 		// is different from requiresValue in that requests that aren't fully
> 		// covered must
> ```

nit: unfinished thought.
___
*[`pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer_test.go` line 2939 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/148438#-OT-XA4w3vbJ8ldnk1aD:-OT-XA4w3vbJ8ldnk1aE:b-fpbtw0) ([raw file](https://github.com/cockroachdb/cockroach/blob/47ec2e06784b20fd12ea6ea77562b4c397e42983/pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer_test.go#L2939)):*
> ```Go
> 				// Encode to BATCH_RESPONSE FORMAT.
> 				pebbleResultsKVLenSize := 8
> 				keyLen, valLen := encKVLength(keyA, &valueA)
> ```

nit: this seems incorrect - the first return value of `encKVLength` is `lenKV` which is the total length of the encoded KV, including the MVCC stuff. In other words, we should do something like:
```
lenKV, _ := encKVLength(keyA, &valueA)
repr := make([]byte, 0, lenKV)
```
(i.e. we currently allocate more than needed. We have some validation for no leftover space in `respMerger.toScanResp` but probably that doesn't get called in this test.)
___
*[`pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer.go` line 769 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/148438#-OT-T0Hl-3u3GzIHAydf:-OT-T0Hl-3u3GzIHAydg:bf52ix2) ([raw file](https://github.com/cockroachdb/cockroach/blob/47ec2e06784b20fd12ea6ea77562b4c397e42983/pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer.go#L769)):*
> ```Go
> 			// If the MustAcquireExclusiveLock flag is set on the Put, then we
> 			// need to add a locking Get to the BatchRequest, including if the
> 			// key doesn't exist.
> ```

nit: let's adjust the comment with new addition.
___
*[`pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer.go` line 815 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/148438#-OT-Uv6r6AeqJr4WPSZ3:-OT-Uv6r6AeqJr4WPSZ4:bekxtk6) ([raw file](https://github.com/cockroachdb/cockroach/blob/47ec2e06784b20fd12ea6ea77562b4c397e42983/pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer.go#L815)):*
> ```Go
> 					// Even though the Get request must be served from the buffer, as the
> 					// transaction performed a previous write to the key, we still need to
> 					// acquire a lock at the leaseholder. As a result, we can't strip the
> ```

nit: let's adjust the comments in both branches of this `if` statement. Also need to adjust the logging in the `else` branch.


<!-- Sent from Reviewable.io -->

stevendanna reviewed (commented):
TFTR! I answered your non-nit questions. I'll push fixes to the nits shortly.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/148438#-:-OT08CKE8mlSk98R8m_5:bidjad7)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @yuzefovich)

___
*[`pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer_test.go` line 2939 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/148438#-OT-XA4w3vbJ8ldnk1aD:-OT0CVreC5ObLKx_M7f_:b-5hynew) ([raw file](https://github.com/cockroachdb/cockroach/blob/47ec2e06784b20fd12ea6ea77562b4c397e42983/pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer_test.go#L2939)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: this seems incorrect - the first return value of `encKVLength` is `lenKV` which is the total length of the encoded KV, including the MVCC stuff. In other words, we should do something like:
```
lenKV, _ := encKVLength(keyA, &valueA)
repr := make([]byte, 0, lenKV)
```
(i.e. we currently allocate more than needed. We have some validation for no leftover space in `respMerger.toScanResp` but probably that doesn't get called in this test.)
</blockquote></details>

Ah, thanks!


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (approved):
The change <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> will defer to you whether to wait for Arul to also take a look.

> Reviewed 2 of 2 files at r4, 2 of 2 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/148438#-:-OT36GoW3oSew8kqzYcs:b-1dbjdz)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @arulajmani and @stevendanna)

___
*[`pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer_test.go` line 2939 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/148438#-OT-XA4w3vbJ8ldnk1aD:-OT34DUcAKS4vbS-WXjS:bgk5o92) ([raw file](https://github.com/cockroachdb/cockroach/blob/47ec2e06784b20fd12ea6ea77562b4c397e42983/pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer_test.go#L2939)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

Ah, thanks!
</blockquote></details>

I don't think we need `pebbleResultsKVLenSize` too (`lenKV` should already include it) - is it necessary?
___
*[`pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer.go` line 791 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/148438#-OT35kgI-D8VzUhiUw4-:-OT35kgI-D8VzUhiUw40:b-bm8ghm) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e96dbbd3bc89623ae0328b81a802ea03fedb729/pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer.go#L791)):*
> ```Go
> 			// need to add a locking Get to the BatchRequest, including if the key
> 			// doesn't exist. We can elide this locking request when we have both a
> 			// value (to populated the FoundKey field in the response) and a lock.
> ```

nit: `s/to populated/to populate/`.


<!-- Sent from Reviewable.io -->
